lib/fauxpaas/artifact.rb

```diff
@@ -2,48 +2,29 @@
 
 module Fauxpaas
   # A set of files on disk as they will be uploaded to an
-  # application server. Pending work on AEIM-1357, AEIM-1361, AEIM-1362,
+  # application server. Pending work on AEIM-1361, AEIM-1362,
   # AEIM-1375, AEIM-1363, etc to fully expose the proper shape of this.
   class Artifact
-    def initialize(signature:, fs:)
-      dir = fs.mktmpdir
-      @fs = fs
+    def initialize(signature:, ref_repo:)
+      dir = Pathname.new(Dir.mktmpdir)
+      @ref_repo = ref_repo
       @shared_path = extract_ref(signature.shared, dir/"shared")
       @unshared_path = extract_ref(signature.unshared, dir/"unshared")
       @source_path = extract_ref(signature.source, dir/"source")
     end
 
-    def extract_ref(ref, path)
-      fs.mkdir_p path
-      add_reference(ref, path)
-      path
-    end
-
-    # This will be refacted in the course of AEIM-1357
-    # rubocop:disable Metrics/AbcSize
-    # rubocop:disable Metrics/MethodLength
-    def add_reference(reference, base)
-      fs.mkdir_p(base)
-      fs.mktmpdir do |dir|
-        reference.checkout(dir) do |working_dir|
-          working_dir
-            .relative_files
-            .reject {|path| fs.directory?(path) }
-            .map {|file| [working_dir.dir/file, base/file] }
-            .each do |src, dest|
-              fs.mkdir_p(dest.dirname)
-              fs.cp(src, dest)
-            end
-        end
-      end
-    end
-    # rubocop:enable Metrics/MethodLength
-    # rubocop:enable Metrics/AbcSize
-
     attr_reader :source_path, :shared_path, :unshared_path
 
     private
 
-    attr_reader :fs
+    attr_reader :ref_repo
+
+    def extract_ref(ref, path)
+      ref_repo.resolve(ref)
+        .cp(path)
+        .write
+        .path
```

Comment:
Seems that this should maybe go to ReferenceRepo (and be called in the initializer). I found it confusing to get a lazy dir from the repo and then work with it here.

Alternatively, the "work" phase for the artifact could initiated by a public method. I tend to prefer potentially long/expensive operations to happen outside of initializers where the API is amenable. I don't mind something like `Artifact.new(details).build!` or if the public readers are guarded by something like `ensure_unpacked!`.

Or maybe this is really hiding the notion that getting a functional Artifact is potentially expensive, complicated, or error prone, and at least a factory method is in order. I just find the side effect of initiating the whole clone/copy process via the constructor a bit dodgy (contrast Net::HTTP, where instantiation, execution, and combo convenience methods are all separately treated).

Reply:
I expect this (and the calls from the initializer) will ultimately go to `ArtifactBuilder`. That change is not in scope for this issue, though.